test/e2e/util_gatewayapi_test.go

```diff
@@ -0,0 +1,611 @@
+		}
+		if catalogSource.Status.GRPCConnectionState != nil && catalogSource.Status.GRPCConnectionState.LastObservedState == "READY" {
+			t.Logf("found catalogSource %s with last observed state %s", catalogSource.Name, catalogSource.Status.GRPCConnectionState.LastObservedState)
+			return true, nil
+		}
+		t.Logf("found catalogSource %s but could not determine last observed state, retrying...", catalogSource.Name)
+		return false, nil
+	})
+	return err
+}
+
+// assertSMCP checks if the ServiceMeshControlPlane exists in a ready state,
+// and returns an error if not.
+func assertSMCP(t *testing.T) error {
+	t.Helper()
+	smcp := &maistrav2.ServiceMeshControlPlane{}
+	nsName := types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: openshiftSMCPName}
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 3*time.Minute, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, nsName, smcp); err != nil {
+			t.Logf("failed to get ServiceMeshControlPlane %s/%s: %v, retrying...", nsName.Namespace, nsName.Name, err)
+			return false, nil
+		}
+		if smcp.Status.Readiness.Components != nil {
+			pending := len(smcp.Status.Readiness.Components["pending"]) > 0
+			unready := len(smcp.Status.Readiness.Components["unready"]) > 0
+			if pending || unready {
+				t.Logf("found ServiceMeshControlPlane %s/%s, but it isn't ready. Retrying...", smcp.Namespace, smcp.Name)
+				return false, nil
+			}
+			if len(smcp.Status.Readiness.Components["ready"]) > 0 {
+				t.Logf("found ServiceMeshControlPlane %s/%s with ready components: %v", smcp.Namespace, smcp.Name, smcp.Status.Readiness.Components["ready"])
+				return true, nil
+			}
+		}
+		t.Logf("found ServiceMeshControlPlane %s/%s but could not determine its readiness. Retrying...", smcp.Namespace, smcp.Name)
+		return false, nil
+	})
+	return err
+}
+
+// assertDNSRecord checks to make sure a DNSRecord exists in a ready state,
+// and returns an error if not.
+func assertDNSRecord(t *testing.T, recordName types.NamespacedName) error {
+	t.Helper()
+	dnsRecord := &v1.DNSRecord{}
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, recordName, dnsRecord); err != nil {
+			t.Logf("failed to get DNSRecord %s/%s: %v, retrying...", recordName.Namespace, recordName.Name, err)
+			return false, nil
+		}
+		// Determine the current state of the DNSRecord.
+		if len(dnsRecord.Status.Zones) > 0 {
+			for _, zone := range dnsRecord.Status.Zones {
+				for _, condition := range zone.Conditions {
+					if condition.Type == v1.DNSRecordPublishedConditionType && condition.Status == string(metav1.ConditionTrue) {
+						return true, nil
+					}
+				}
+			}
+		} else {
+			t.Logf("found DNSRecord %s/%s but could not determine its readiness. Retrying...", recordName.Namespace, recordName.Name)
+			return false, nil
+		}
```

Comment:
*nit* It only logs when `dnsRecord.Status.Zones` == 0, but if you didn't find the condition, it doesn't log. Seems like you could simplify and allow the log to apply to both.

```suggestion
		}
		t.Logf("found DNSRecord %s/%s but could not determine its readiness. Retrying...", recordName.Namespace, recordName.Name)
```